A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,529 @@
+NACK from client | false | (any) | `OnError(status)`, Fail data plane RPCs | `OnResourceChanged(Status)`, Fail data plane RPCs
+NACK from client | true | false | `OnError(status)`, Ignore | `OnAmbientError(Status)`, Ignore
+NACK from client | true | true  | `OnError(status)`, Ignore | `OnResourceChanged(Status)`, Drop existing resource and fail RPCs
+Resource timeout | false | (any) | `OnResourceDoesNotExist()`, Fail data plane RPCs | `OnResourceChanged(Status(NOT_FOUND))` (status will be UNAVAILABLE if `resource_timer_is_transient_error` server feature is present), Fail data plane RPCs
+LDS or CDS resource deletion from server | true | false | `OnResourceDoesNotExist()`, Drop existing resource and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnAmbientError(Status(NOT_FOUND))`, Ignore
+LDS or CDS resource deletion from server | true | true | `OnResourceDoesNotExist()`, Drop existing resources and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnResourceChanged(Status(NOT_FOUND))`, Drop existing resource and fail RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | false | N/A | `OnAmbientError(status)`, Ignore
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | true | N/A | `OnResourceChanged(status)`, Drop existing resource and fail RPCs
+[xRFC TP3] error with other status | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with other status | true | (any) | N/A | `OnAmbientError(status)`, Ignore
+
+### Handling Ambient Errors
+
+As mentioned above, the purpose of ambient errors is to provide useful
+context in error messages that are triggered for reasons not directly
+related to the ambient error.  For example, if we lose contact with the
+xDS server, we continue to use cached resources, but if we subsequently
+lose contact with all endpoints, we will fail data plane RPCs.  The RPC
+failure is triggered by the fact that we've lost contact with all
+endpoints, but the status message should include the fact that we've
+lost contact with the xDS server, since the outage may be caused by the
+fact that we haven't been getting EDS updates from the xDS server.
+
+To support this, we will change pick_first and the petiole policies (see
+[gRFC A61]) to include the `resolution_note` field in all error messages
+generated by those policies.  The xds resolver will set the resolution
+note in the `XdsConfig` (see [gRFC A74]) to include ambient errors for
+all resources (LDS, RDS, CDS, and EDS).  If none of the resources
+have ambient errors, the xds resolver will set the resolution note to
+indicate the xDS node ID, which can be useful for debugging.
+
+### Propagating Status to Data Plane RPCs
```

Comment:
Maybe forward-ref this section above in at least one place where you talk about failing data plane RPCs?

Reply:
I like this idea, but I don't see an obvious place where it makes sense to do so where it wouldn't seem cumbersome.

FWIW, this proposal isn't really changing anything from A54 at all here; the only reason I added this section was to make sure implementors keep this in mind and don't accidentally break it when implementing the new watcher API.